### PR TITLE
Automatically save and allow restore of recent DAG run configs

### LINKED ISSUE #27714: Re-use recent DagRun JSON-configurations
### Description

Allow users to re-use recent DagRun configurations upon running a DAG.
This can be achieved by adding a dropdown that contains some information about recent configurations. When user selects an item, the relevant JSON configuration can be pasted to the "Configuration JSON" textbox.
Screen Shot 2022-11-16 at 16 22 30
Screen Shot 2022-11-16 at 16 22 38
Screen Shot 2022-11-16 at 16 38 40

### Use case/motivation

Commonly, DAGs are triggered using repetitive configurations. Sometimes the same configuration is used for triggering a DAG, and sometimes, the configuration differs by just a few parameters. 

This interaction forces a user to store the templates he uses somewhere on his machine or to start searching for the configuration he needs in `dagrun/list/`, which does take extra time.

It will be handy to offer a user an option to select one of the recent configurations upon running a DAG.

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by aaronabraham311**
closes: #27714

Note: we found that the restored recent config isn't automatically formatted because CodeMirror doesn't support that anymore. We could add js-beautify but we were unsure if we should add this library for this singular fix. It would also only work for ES10. If we want more backwards compatability, let us know and we can re-write.

cc: @jh242

### CODE REVIEW

**Comment by ephraimbuddy** [comment]
If you don't mind, you can add before and after pics

**Comment by uranusjr** [comment]
What happens if the previous config is big? I wonder if we should simply show a datetime or sorts for each saved entry instead.

Also do we need a way for the user to clear saved configs?

**Comment by pierrejeambrun** [comment]
Is it relevent to see 'previous config' of all other dag ? Should this be on a per dag basis ?

**Comment by pierrejeambrun** [comment]
Also wondering if the localstorage is the correct place for this. Storing all configs of each manual run for each dag could be pretty heavy. (localstorage is limited to 5M). Also loading all of this on the page could be quite long (storage is synchronous if i am not mistaken so ready all of that from it could be expensive on a UI point of view)

**Comment by aaronabraham311** [comment]
> Also wondering if the localstorage is the correct place for this. Storing all configs of each manual run for each dag could be pretty heavy. (localstorage is limited to 5M). Also loading all of this on the page could be quite long (storage is synchronous if i am not mistaken so ready all of that from it could be expensive on a UI point of view)

@pierrejeambrun Do you have an alternative suggestion of storing recent configs asides from localstorage?

**Comment by pierrejeambrun** [comment]
@aaronabraham311, one other obvious way would be to persist them in the db, and expose them through the rest api.

This requires more effort to implement, and I am not sure all db support json type. This is just an idea, wondering what others think.

**Comment by jh242** [comment]
I disagree with the db/API endpoint idea. I think it's too much overhead for a minor feature that seems intended to save some time for DAGs with small configs that the user didn't expect to run multiple times. I feel like the ability to copy/paste configurations from [here](https://github.com/apache/airflow/pull/27639) also overlaps this feature. My suggestion is that we can either save recent configs in session storage by DAG, or in local storage but limited to a certain number of total recent configs across all DAGs.

Additionally, Aaron and I are on a bit of a short timeline and likely won't have time to implement a backend-supported version of this feature, but if that's the direction we really want to go, we can get started on it and see where it goes.

**Comment by pierrejeambrun** [comment]
Now that you mention it, DagRun conf are already stored and available for each run. Isn't it easier to just retrieve them and provide them to the `trigger.html` template directly ? We also have a lot of control of what conf should be retrieved (5 most recent conf for a specific dag, etc.)

**Comment by aaronabraham311** [comment]
@pierrejeambrun Oh that's great! Is there any example on how to access the DagRun configs from the db? Or is there a code snippet that we can use as an example?

**Comment by pierrejeambrun** [comment]
@aaronabraham311 There is a lot of example querying resource from the db in the views.py file. In this case DagRun should have what you need. In the example you mentioned above (json config displayed in the dagrun details), its coming from the grid_data view of that file :)

**Comment by jh242** [comment]
> @aaronabraham311 There are a lot of examples querying resource from the db in the [`views.py`](https://github.com/apache/airflow/blob/main/airflow/www/views.py) file. In this case DagRun should have what you need. In the example you mentioned above (json config displayed in the dagrun details), its coming from the [`grid_data`](https://github.com/apache/airflow/blob/main/airflow/www/views.py#L3487) view of the same file :)

I've implemented this approach and limited the number of requested recent DAG runs to 10... Because there may be duplicate recent configs this means we won't necessarily get 10 different recent configs. Take a look!

**Comment by pierrejeambrun** [comment]
Hello @jh242,

Thank you for the update. Here you are loading the grid_data info (which has a lot of extra stuff on task instances etc.), just to retrieve the 'configs'. As you said you could end up with less than 10 confs (because of duplicate), or even with no conf at all if the last 10 run were scheduled.

Here we statically render a template, there is usually no need to go through an API call. What you could do is, inside the view that renders the `trigger.html`, make your custom request to the db to retrieve the conf you like. (Filter on confs that are not of type json, deduplicate them with a distrinc(), fetch only the last runs that were 'manually triggered', with a conf `not None`) etc. When you have your configurations, you can pass this as an extra variable to the `trigger.html` template. Then update the template so it can use this variable to render extra options in the drop down.

You can check the `def index` (/home) view and check how we directly pass variable to the template for rendering. (dags, tags etc). 

Maybe I missed something, but this should make the code even cleaner and simple :)

**Comment by jh242** [comment]
Thanks @pierrejeambrun , that's an excellent point. My latest commit implements your suggestion.

File: airflow/www/templates/airflow/trigger.html
Comment (pierrejeambrun) [inline airflow/www/templates/airflow/trigger.html]: Maybe we can display something like 'empty config' instead of None that doesn't provide much information to the user ?

File: airflow/www/templates/airflow/trigger.html
Comment (pierrejeambrun) [inline airflow/www/templates/airflow/trigger.html:63]: This div was here to add some spacing after the json editor. (It wasn't working properly indeed). Maybe adding a line break (<br>), currently it's quite compact :):
image

File: airflow/www/static/js/trigger.js
Comment (pierrejeambrun) [inline airflow/www/static/js/trigger.js]: To avoid manually replacing the quotes, you can `json.dumps(conf)`. See what is done in the grid_data endpoint, and my comments bellow.

File: airflow/www/views.py
Comment (pierrejeambrun) [inline airflow/www/views.py]: I don't think empty objects are relevent here. (also update the `is not None` syntax)
```suggestion
                DagRun.conf.isnot(None),
                DagRun.conf != {}
```

File: airflow/www/views.py
Comment (pierrejeambrun) [inline airflow/www/views.py]: Conf are not always json compatible. I believe they can be plain `str`. check `def encode_dag_run` func. We may want to also handle this case. Maybe we can extract a reusable function that will handle the conf the same way that is done in `encode_dag_run`. (`json.dumps` if possible, otherwise `str`, and also returns `conf_is_json`, then we skip the conf that are not json in the response)

**Review by pierrejeambrun — COMMENTED** [review]
Looking good, tested locally and seems great.

I made a few suggestions. Can you also rebase your branch please ? :)

@bbovenzi do you mind providing some feedback on this one, as you already handled confs in the `grid_data` view. :)

**Comment by jh242** [comment]
Hey @pierrejeambrun , thanks so much for your guidance so far. I've implemented your suggestions, including the reusable function for checking if dagRun configs are JSON or not. I'm still a little bit confused on serializing the response from my query as I haven't worked with Flask before and I wasn't sure how to test a string dagRun config, so that may need more attention.

**Comment by aaronabraham311** [comment]
Hi @pierrejeambrun @bbovenzi! We have addressed your requested changes. Please let us know if we can begin merging this! We would like to merge this in by end of this week if possible as we won't be able to push this through afterwards.

File: airflow/www/utils.py
Comment (pierrejeambrun) [inline airflow/www/utils.py]: return type

**Review by pierrejeambrun — COMMENTED** [review]
Looking good to me. I would like to get a second pair of eyes on this one. A few nits on my side and I should be ready :)

File: airflow/www/utils.py
Comment (jh242) [reply airflow/www/utils.py]: done!

File: airflow/www/static/js/trigger.js
Comment (bbovenzi) [inline airflow/www/static/js/trigger.js]: I wonder if we should try to format the config JSON so it doesn't appear as a single line

```
function setRecentConfig(e) {
  let { value } = e.target;
  try {
    const json = JSON.parse(value);
    value = JSON.stringify(json, null, 2);
  } catch (err) {
    console.err(err);
  }
  document.querySelector('.CodeMirror').CodeMirror.setValue(value);
}
```

File: airflow/www/views.py
Comment (bbovenzi) [inline airflow/www/views.py:1908]: Can we change this to the first 5 unique configs?
If we can now select old configs, a lot of the options would just be the same.

File: airflow/www/static/js/trigger.js
Comment (jh242) [reply airflow/www/static/js/trigger.js]: Hey yeah! I wanted to do this as well, but code formatting is a function that CodeMirror no longer supports. The other alternative is adding another JS library just to do this formatting, but I don't think we need this overhead (let me know if we already have one that can do it!).

The other alternative is writing a bespoke function to do JSON formatting but I didn't think it was worth the trouble.

File: airflow/www/views.py
Comment (jh242) [reply airflow/www/views.py:1908]: Whoops, yep! Somehow that slipped my mind.

File: airflow/www/static/js/trigger.js
Comment (bbovenzi) [reply airflow/www/static/js/trigger.js]: The function above should do it. We just need that try/catch block to capture thrown errors when it is not valid json.

File: airflow/www/views.py
Comment (jh242) [reply airflow/www/views.py:1908]: I'm a bit of an SQLalchemy noob--I thought that dropping in a .distinct after (or before) the .order_by clause would produce what we needed, but it doesn't seem to be working and I'm not sure why. Could I get a bit of guidance on this @bbovenzi ? Thanks in advance!

Comment (bbovenzi) [reply airflow/www/views.py:1908]: I think `.distinct(DagRun.conf)` works?

Comment (jh242) [reply airflow/www/views.py:1908]: ~~I've tried this and it did not. Removing the .order_by made it work though.~~ Works now!

**Comment by pierrejeambrun** [comment]
@jh242 Can you also fix the failing text. The extracted function `get_dag_run_conf` is not exactly doing what was done before directly in the `encode_dag_run`. I recommend keeping the same structure.

(There is a difference if the dagrun is an empty object)

**Comment by jh242** [comment]
@pierrejeambrun @bbovenzi My latest commit should address all of the above issues! Thanks so much again guys. Would really appreciate getting this merged before Friday if possible--we're contributing as part of a university course and it would be great to have something to show for our efforts :D

File: airflow/www/static/js/trigger.js
Comment (bbovenzi) [inline airflow/www/static/js/trigger.js]: We can throw a `console.error()` here, but lets leave the value as a single string if we can't parse it.

Comment (jh242) [reply airflow/www/static/js/trigger.js]: done

**Review by bbovenzi — APPROVED** [review]
Nice work. Thanks for being patient with all the requested changes.

**Comment by jh242** [comment]
updated our branch to be up-to-date with airflow. Thanks again @bbovenzi @pierrejeambrun :D

**Review by pierrejeambrun — APPROVED** [review]
LGTM ! Thank you for this addition :tada:

**Pull request merged**
